File: frontend/src/types/redlines.ts
/**
 * Types mirroring the suggested-redline backend.
 *
 * Wire shape matches `backend/app/schemas/redlines.py`. Each redline
 * is a non-deterministic LLM-generated suggestion attached to a
 * failed `DeviationFinding`. The status workflow is independent of
 * the parent finding's `finding_status` — a finding can be marked
 * `reviewed` while one of its redlines is still `proposed`.
 */
export type RedlineStatus = "proposed" | "accepted" | "rejected";

export interface SuggestedRedline {
  id: string;
  organization_id: string;
  contract_id: string;
  finding_id: string;
  redline_text: string;
  rationale: string | null;
  model_name: string;
  prompt_version: string;
  confidence: number;
  status: RedlineStatus;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface UpdateRedlineStatusRequest {
  status: RedlineStatus;
}

File: backend/app/schemas/redlines.py
"""Request and response schemas for LLM-generated suggested redlines.

Wire shape mirrors the persisted ``SuggestedRedline`` ORM row.
Reviewer-mutable fields (``status`` only) are validated by the
``UpdateRedlineStatusRequest`` body schema; everything else is
immutable post-write.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict

# Reviewer-settable values for the redline workflow. Every value the
# backend persists is also valid here — there is no equivalent of the
# findings ``superseded`` state that the API hides.
RedlineStatus = Literal["proposed", "accepted", "rejected"]


class UpdateRedlineStatusRequest(BaseModel):
    """Body for ``PATCH /api/contracts/{contract_id}/findings/{finding_id}/redlines/{redline_id}``."""

    model_config = ConfigDict(extra="forbid")

    status: RedlineStatus


class SuggestedRedlineResponse(BaseModel):
    """Public projection of one persisted ``SuggestedRedline`` row."""

    model_config = ConfigDict(extra="forbid")

    id: uuid.UUID
    organization_id: uuid.UUID
    contract_id: uuid.UUID
    finding_id: uuid.UUID
    redline_text: str
    rationale: str | None = None
    model_name: str
    prompt_version: str
    confidence: float
    status: RedlineStatus
    created_by: uuid.UUID | None = None
    created_at: datetime
    updated_at: datetime
